lib/pos_fee_estimator.go

```diff
@@ -0,0 +1,506 @@
+		}
+	}
+	return nil
+}
+
+// addBlockToTransactionRegister adds all the transactions from the block to the given transaction register.
+// Should only be called when the rwLock over a TransactionRegister is held.
+func addBlockToTransactionRegister(txnRegister *TransactionRegister, block *MsgDeSoBlock) error {
+	for _, txn := range block.Txns {
+		// We explicitly exclude block reward transactions as they do not have fees.
+		if txn.TxnMeta.GetTxnType() == TxnTypeBlockReward {
+			continue
+		}
+		mtxn, err := NewMempoolTx(txn, block.Header.TstampNanoSecs, block.Header.Height)
+		if err != nil {
+			return errors.Wrap(err, "PoSFeeEstimator.addBlockToTransactionRegister: error creating MempoolTx")
+		}
+		if err = txnRegister.AddTransaction(mtxn); err != nil {
+			return errors.Wrap(err,
+				"PoSFeeEstimator.addBlockToTransactionRegister: error adding txn to pastBlocksTransactionRegister")
+		}
+	}
+	return nil
+}
+
+// RemoveBlock removes a block from the PoSFeeEstimator. This will remove all the transactions from the block
+// from the pastBlocksTransactionRegister and remove the block from the cache.
+func (posFeeEstimator *PoSFeeEstimator) RemoveBlock(block *MsgDeSoBlock) error {
+	posFeeEstimator.rwLock.Lock()
+	defer posFeeEstimator.rwLock.Unlock()
+
+	if err := posFeeEstimator.removeBlockNoLock(block); err != nil {
+		return errors.Wrap(err, "PoSFeeEstimator.RemoveBlock: error removing block from PoSFeeEstimator")
+	}
+	return nil
+}
+
+// removeBlockNoLock is the same as RemoveBlock but assumes the caller has already acquired the rwLock.
+func (posFeeEstimator *PoSFeeEstimator) removeBlockNoLock(block *MsgDeSoBlock) error {
+	// Remove all transaction from the block from the pastBlocksTransactionRegister.
+	for _, txn := range block.Txns {
+		// We explicitly exclude block reward transactions as they do not have fees.
+		// They were never added in the first place.
+		if txn.TxnMeta.GetTxnType() == TxnTypeBlockReward {
+			continue
+		}
+		mtxn, err := NewMempoolTx(txn, block.Header.TstampNanoSecs, block.Header.Height)
+		if err != nil {
+			return errors.Wrap(err, "PoSFeeEstimator.RemoveBlock: error creating MempoolTx")
+		}
+		if err = posFeeEstimator.pastBlocksTransactionRegister.RemoveTransaction(mtxn); err != nil {
+			return errors.Wrap(err,
+				"PoSFeeEstimator.removeBlockNoLock: error removing txn from pastBlocksTransactionRegister")
+		}
+	}
+	blockHash, err := block.Hash()
+	if err != nil {
+		return errors.Wrap(err, "PoSFeeEstimator.removeBlockNoLock: error computing blockHash")
+	}
+	// Remove the block from the cached blocks.
+	for ii, cachedBlock := range posFeeEstimator.cachedBlocks {
+		cachedBlockHash, err := cachedBlock.Hash()
+		if err != nil {
+			return errors.Wrap(err, "PoSFeeEstimator.removeBlockNoLock: error computing cachedBlockHash")
+		}
+		if blockHash.IsEqual(cachedBlockHash) {
+			posFeeEstimator.cachedBlocks = append(posFeeEstimator.cachedBlocks[:ii],
```

Comment:
A long time ago I hit a bug related to reusing a slice in an append like this. Something about rewriting the slice in-place. I'm not sure if you have a bug here because of it, but the safer thing would be to make a copy of cachedBlocks like this:

```
newCachedBlocks = []
for ii, cachedBlock := range posFeeEstimator.cachedBlocks {
    if blockHash.IsEqual(cachedBlockHash) {
        continue
    }
    newCachedBlocks = append(newCachedBlocks, cachedBlock)
}
posFeeEstimator.cachedBlocks = newCachedBlocks
```